### DISCUSSION TITLE
'Removal' Operation to the Cart Transformation API for Enhanced Flexibility

### DISCUSSION THREAD

**Discussion by 0xtlt** [Shopify/function-examples#213] [Feedback: Cart Transform API]
Hi everyone!

The Cart Transformation API is going to be a fantastic addition!

As of now, this API offers 2 "operations":
- expand
- merge

I believe it would be beneficial to introduce an additional operation, a "removal" operation, which enables the removal of a product/variant from the cart.

This operation could potentially:
- Prevent a customer from purchasing the same product more than once (using a tag added through a flow).
- Prevent simultaneous purchases of two products that are not meant to be bought together.

The proposed `DeleteOperation` would include:
- cartLineId of type `ID!`
- quantity of type `Int`, optional (defaulting to the current line quantity)

Alternatively, we could consider implementing a more versatile `EditOperation` or an `AdjustQuantityOperation` that would allow for adjusting line item quantities.

In any case, this is a concept worth delving into. I'm eager to hear your thoughts and engage in a discussion to determine the optimal solution! :)

**Comment by gil--** [comment]
One other common use-case for this would be auto removal of a free gift if requirements are no longer met such as falling under a required subtotal. I am pretty sure this was possible via scripts.

**Reply by ignacio-chiazzo** [reply]
Having cases like this is very helpful to prioritize. Please keep sharing them :)

**Reply by cwveysey** [reply]
@ignacio-chiazzo @dianacheptene good afternoon! I hope that this finds you as well as possible.

When convenient, can you please confirm - 

Has something similar to the "DeleteOperation" described [here](https://github.com/Shopify/function-examples/discussions/213#discussion-5071923) been added to the roadmap? Something that for example would help merchants facilitate [this](https://github.com/Shopify/function-examples/discussions/213#discussioncomment-5581970).

Copying @dianacheptene because she was mentioned elsewhere on this page.

Thank you!

**Comment by adityaagashe1** [comment]
There's another common use-case that we can handle this with forcing max-qty on a lineitem. If user order has more than allowable max quantity it would be nicer to remove it automatically.

**Reply by CodeBrotha** [reply]
Yes! The lack of this functionality has prevented my team from moving off Shopify Scripts.

**Comment by ignacio-chiazzo** [comment]
Hey! thanks for the feedback.

Could you elaborate more on why not using the Cart API to apply Cart Lines changes?

**Reply by gil--** [reply]
With checkout extensions, it's possible someone adds more quantity than allowed. The upcoming cart validations allow for blocking checkout if say a gift is in the cart but the requirements are not met. This is OK but not ideal since often the merchant wants to simply auto-remove the free gift. Relying on an extension to do this (auto remove) is not ideal and feels like something better done with functions.

**Reply by bkspace** [reply]
I think this is more important when it comes to Storefront functionality, but also not ideal in the Checkout. We currently use an App Theme extension to detect when to auto add or auto remove a product. However, there is no common way/interface to 'refresh' a cart across different themes. A function to add/remove a product would bypass this issue as the logic can be run prior to the cart being returned to the theme.

**Reply by sadsciencee** [reply]
ignacio -

for public apps on stores that don't use shopify plus, we are not able to use the cart api at checkout

additionally, assuming that one day cart transforms will be more usable on liquid storefronts, the complexity of maintaining apps across 100s of themes would be significantly reduced if we could run certain cart operations server side (add items, remove items, edit properties)

**Comment by rbarreca** [comment]
We also need this and would prefer a more versatile `EditOperation` or `AdjustQuantityOperation`. 

Our use case: we sell perishable food for local delivery (e.g. sprouts with very short shelf life). Grocery delivery is different than most other forms of e-commerce in that the customer generally selects the delivery date first, then adds items to their cart _that they expect to receive on that day_. Our customers set their delivery date while shopping (as a cart attribute) and then we subvert Shopify's native concept of inventory and instead display each product variant's inventory by looking up a product variant metafield hash (an object describing 3 sprouts avail tmrw, 0 avail Sat, 12 avail Sun, as `{ "2023-11-03": 3, "2023-11-04": 0, "2023-11-05": 12 }`). 

A customer may select delivery for tomorrow, add 3 sprouts to cart, then change delivery date to Saturday. We want to then remove those items from the cart via a function, but currently have no way to do so. We've built a workaround using theme liquid, cart AJAX API, and a checkout UI extension calling `applyCartLinesChange`. But, it requires us to maintain code in multiple places and not agnostic to sales channel. We'd rather do this server side in a function so it's faster and less brittle and better to maintain.

Ideally, from a function we could get cart attributes (for the delivery date), each cart line item variant's metafields (for date-based inventory availability), then call an `EditOperation` or `AdjustQuantityOperation` to decrement quantities of items that are no longer available for the chosen delivery date (either because the user changed date at the last minute or someone else swooped in and purchased those item quantities before the user had a chance to checkout).

**Comment by iska31** [comment]
Hello @dianacheptene ,

Do you have any information about an additional "edit" operation for the CartTransform : https://shopify.dev/docs/api/functions/reference/cart-transform/graphql/common-objects/cartoperation ? 

This feature is for some Shopify Plus a must have to fully complete the Checkout extensibility migration.
Thank you,
Best regards

**Reply by gil--** [reply]
What would edit do that update does not currently? https://shopify.dev/docs/api/functions/reference/cart-transform/graphql/common-objects/updateoperation

**Reply by CodeBrotha** [reply]
@gil-- Even this updateOperation is too limited to be useful.

For example, line item quantity is not available. 

I can't think of any common use case for overriding image or title in checkout, but definitely have a few use cases for quantity.

**Reply by iska31** [reply]
@gil--

I need to adjust line item quantities for various reasons:

- To remove unauthorized items, like free samples added to the cart without meeting the necessary conditions.
- To enforce maximum quantity limits per order for certain items.
- To comply with complex quantity conditions.

So far, I haven't needed to change titles, images, or prices (as product discounts often cover price adjustments) of a line item.

While client-side adjustments work for simple scenarios, more intricate purchasing processes require backend logic to ensure all conditions are met. This is increasingly important for merchants aiming to offer unique shopping experiences, especially when paying the Shopify Plus solution.

**Comment by jo-elbdev** [comment]
This feature is still very much needed. Without this, multiple online shops cannot migrate from Shopify Scripts to the Shopify Functions. Did someone find any workaround here?

**Reply by CodeBrotha** [reply]
Is there any update to this issue. The deadline for checkout extensibility is approaching and this issue is still preventing us from abandoning Shopify Scripts.

**Comment by msev** [comment]
Any news from Shopify on this subject?

We still have to use Shopify Scripts as it is, because it's not possible to delete line items or adjust quantity via Shopify Functions.

**Reply by gil--** [reply]
Just a reminder that you can leverage checkout ui extensions to adjust line items: https://shopify.dev/docs/api/checkout-ui-extensions/2023-10/apis/cart-lines#checkoutapi-propertydetail-applycartlineschange 

At this time, I would recommend using that API for anything that is modifying the cart lines.

**Reply by msev** [reply]
Thanks for your answer @gil-- 

We actually tested this but it's not the equivalent of Shopify Scripts.

You can only modify one line item at a time (not several in a single request).
Shopify Scripts runs on the server side when [checkout ui extensions runs on the client side](https://shopify.dev/docs/apps/build/checkout/cart-checkout-validation#client-side-validation) and only in the checkout.

Shopify Scripts runs on the cart in global (theme/checkout), and therefore avoids having to do the equivalent of the JavaScript mechanics on the theme side, which won't be the case with checkout ui extensions.

We'd need the equivalent of Shopify Scripts in Shopify Functions.

**Reply by Rabot18** [reply]
@gil-- yes it's the only solution for now but it's not server side so it can be bypassed... With Shopify Scripts it's on server side.

**Reply by gil--** [reply]
Thanks for the additional context!

**Comment by mubbo** [comment]
We have a merchant that offers an assembly service for childrens furniture, someone comes to the house and assembles the furniture for them.  So if someone adds a assembly product to the cart and then decides at some point before heading to the checkout to remove the product or last product that requires assembly, the asembly product should automatically be removed from the cart.  

Since I am using the function to modify the costs of the assembly based on metafield information on the product requiring assembly, it would be benificial to also be able to remove the product at the same time instead of having to also add javascript to the theme and monitor what products are in the cart.

**Comment by ezmiller** [comment]
This would be a useful feature for us as well, related to adding/removing gift wrap items.

**Comment by Tartofraise** [comment]
This would be very helpful to create bundle automatically from item of the cart. With parent item and child items being already in the card. If i use expand it will make duplicated child that i have to remove.

**Comment by yashsony** [comment]
Any update here ?

**Comment by alistairmartin** [comment]
Need this feature! Need to remove free gifts if a customer has already redeemed for a promotion. 1 Per Customer per Promotion.
